lib/IRGen/GenArchetype.cpp

```diff
@@ -370,22 +155,107 @@ llvm::Value *irgen::emitArchetypeWitnessTableRef(IRGenFunction &IGF,
   // a cache entry for it) or there's an associated type declaration with
   // that protocol listed as a direct requirement.
 
+  auto localDataKind =
+    LocalTypeDataKind::forAbstractProtocolWitnessTable(protocol);
+
   // Check immediately for an existing cache entry.
-  auto wtable = IGF.tryGetLocalTypeData(archetype,
-                  LocalTypeDataKind::forAbstractProtocolWitnessTable(proto));
+  // TODO: don't give this absolute precedence over other access paths.
+  auto wtable = IGF.tryGetLocalTypeData(archetype, localDataKind);
   if (wtable) return wtable;
 
-  // Otherwise, find the best path from one of the protocols directly
-  // conformed to by the protocol, then get that conformance.
+  // It can happen with class constraints that Sema will consider a
```

Comment:
I absolutely agree.  That was existing code that I didn't feel like testing whether it was still necessary.